R/app-reporting.R

```diff
@@ -82,3 +82,45 @@ show_details.list <- function(x) {
   })
   renderTable(dat, colnames = FALSE)
 }
+
+
+#' Create a modal dialog if user is not in required team(s)
+#'
+#' Takes the output from [check_team_membership()] and, if the user is not in
+#' the required teams, creates a modal dialog indicating which teams they need
+#' to belong to and how to request access.
+#'
+#' @param result Output from [check_team_membership()]
+report_missing_membership <- function(result) {
+  UseMethod("report_missing_membership", result)
+}
+
+#' @rdname report_missing_membership
+report_missing_membership.default <- function(result) {
```

Comment:
I'm not yet very familiar with S3 objects - am I understanding correctly that `check_team_membership()` returns an object with a class of `check_pass` or `check_fail`, `report_missing_membership()` sticks to a default output unless it sees a `check_fail` class? I'm curious to learn in what cases you choose to implement this object type.

Reply:
Yup that's right. Now that I'm looking at it again I think this might be overkill, so I pushed an update. A better example of a time when it's a good idea to use S3 is the `check_annotation_keys()` function which does slightly different things for data frames, Synapse files, and file views. If you don't write them as separate methods, then you end up with really long and complex if statements to ensure that the right thing happens for the right type. Another time that S3 methods come in very handy is when you want to define a new method for a commonly used function that you don't control (like writing a `print()` method for a new object type you've created).